src/main/java/org/candlepin/pki/certs/ContentAccessPayloadKeyBuilder.java

```diff
@@ -0,0 +1,123 @@
+    private static final HexFormat HEX_FORMATTER = HexFormat.of();
+
+    private Consumer consumer;
+    private List<Environment> environments = List.of();
+
+    /**
+     * Creates a new content access payload builder instance for the provided consumer.
+     *
+     * @param consumer
+     *  the consumer to build a content access payload builder instance for
+     *
+     * @throws IllegalArgumentException
+     *  if the provided consumer is null
+     */
+    public ContentAccessPayloadKeyBuilder(Consumer consumer) {
+        if (consumer == null) {
+            throw new IllegalArgumentException("consumer is null");
+        }
+
+        this.consumer = consumer;
+    }
+
+    /**
+     * Sets the environments used to generate the content access payload key. If a null value is provided then
+     * no environments will be used.
+     *
+     * @param environments
+     *  the environments used to generate the content access payload key
+     *
+     * @return this instance of ContentAccessPayloadKeyBuilder
+     */
+    public ContentAccessPayloadKeyBuilder setEnvironments(List<Environment> environments) {
+        this.environments = environments == null ? List.of() : environments;
+        return this;
+    }
+
+    /**
+     * Generates and returns a key for a {@link ContentAccessPayload}.
+     *
+     * @return a generated content access payload key
+     */
+    public String build() {
+        String input = new StringBuilder("arches:")
+            .append(getNormalizedArchitectures(consumer))
+            .append("-environments:")
+            .append(getNormalizedEnvironments(environments))
+            .toString();
+
+        try {
+            byte[] hash = MessageDigest.getInstance(HASH_ALGORITHM)
+                .digest(input.getBytes(StandardCharsets.UTF_8));
+
+            return new StringBuilder("v")
+                .append(SCA_CONTENT_ACCESS_PAYLOAD_KEY_VERSION)
+                .append(":")
+                .append(HEX_FORMATTER.formatHex(hash))
+                .toString();
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String getNormalizedArchitectures(Consumer consumer) {
+        SortedSet<String> sorted = new TreeSet<>();
+
+        String arch = consumer.getFact(Consumer.Facts.ARCHITECTURE);
+        if (arch != null) {
+            sorted.add(arch.strip().toLowerCase());
+        }
+
+        String supportedArches = consumer.getFact(Consumer.Facts.SUPPORTED_ARCHITECTURES);
+        if (supportedArches != null) {
+            List<String> sarches = Util.toList(supportedArches.strip().toLowerCase());
+            sorted.addAll(sarches);
+        }
+
+        return String.join(",", sorted);
+    }
+
+    private String getNormalizedEnvironments(List<Environment> environments) {
+        return environments.stream()
+            .map(Environment::getId)
+            .collect(Collectors.joining(","));
```

Comment:
Instead of relying on the callers to give us an ordered environment list, can we instead make sure this method orders them in a consistent manner? It could be in the same way the SQL in the environmentcurator does it (ascending by env id)

Reply:
We can't at the moment. Due to the fact that priority is buried deep in the consumer object and on the table itself, we don't have any way of sharing that once we pass it around. Or, rather, we *do* have a way, but this is it. We would need a custom wrapper object to enforce it.